auth/oidc.go

```diff
@@ -141,65 +174,58 @@ func (op *OIDCProvider) InitOIDCClient() error {
 		return base.RedactErrorf("Issuer not defined for OpenID Connect provider %+v", base.UD(op))
 	}
 
-	config, shouldSyncConfig, err := op.DiscoverConfig()
-	if err != nil || config == nil {
+	ctx := context.Background()
+	provider, _, err := op.DiscoverConfig(ctx)
+	if err != nil || provider == nil {
 		return pkgerrors.Wrap(err, "unable to discover config")
 	}
 
-	clientCredentials := oidc.ClientCredentials{
-		ID: *op.ClientID,
-	}
-	if op.ValidationKey != nil {
-		clientCredentials.Secret = *op.ValidationKey
+	config := oauth2.Config{
+		ClientID:    *op.ClientID,
+		Endpoint:    provider.Endpoint(),
+		RedirectURL: *op.CallbackURL,
 	}
 
-	clientConfig := oidc.ClientConfig{
-		ProviderConfig: *config,
-		Credentials:    clientCredentials,
-		RedirectURL:    *op.CallbackURL,
+	if op.ValidationKey != nil {
+		config.ClientSecret = *op.ValidationKey
 	}
 
 	if op.Scope != nil || len(op.Scope) > 0 {
-		clientConfig.Scope = op.Scope
+		config.Scopes = op.Scope
 	} else {
-		clientConfig.Scope = []string{"openid", "email"}
+		config.Scopes = []string{oidc.ScopeOpenID, "email"}
 	}
 
-	op.OIDCClient, err = oidc.NewClient(clientConfig)
-	if err != nil {
+	// Initialize the prefix for users created for this provider
+	if err = op.InitUserPrefix(); err != nil {
 		return err
 	}
 
-	// Start process for ongoing sync of the provider config
-	if shouldSyncConfig {
-		op.OIDCClient.SyncProviderConfig(op.Issuer)
-	} else {
-		base.Infof(base.KeyAuth, "Not synchronizing provider config for issuer %s...", base.UD(op.Issuer))
-	}
+	idTokenVerifier := provider.Verifier(&oidc.Config{ClientID: *op.ClientID})
 
-	// Initialize the prefix for users created for this provider
-	if err = op.InitUserPrefix(); err != nil {
-		return err
+	op.OIDCClient = &OIDCClient{
+		Provider:        provider,
+		Config:          config,
+		Context:         ctx,
+		IDTokenVerifier: idTokenVerifier,
 	}
 
 	return nil
 }
 
-func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSync bool, err error) {
-
+func (op *OIDCProvider) DiscoverConfig(ctx context.Context) (config *oidc.Provider, shouldSync bool, err error) {
```

Comment:
It looks like we're only ever passing in the background context here.  If that's the case, what's the value in using it here, and in FetchCustomProviderConfig?

Reply:
'the background context per client' doesn't make sense.  There's only a single background context, and it doesn't actually provide any context-type functionality (never cancelled, no deadline, etc). 

In particular, since there's one long-lived OIDCClient per provider that is *not* associated with a given auth request, I don't think it makes sense for OIDCClient to have a Context property.